### PR TITLE
Added re-authentication when the connection to the VC or ESX is down …

### LINKED ISSUE #2835: Can't run docker container or pull new images: Message:NotAuthenticated
Could be some issue with configuration. But this was working for me like 5 mins ago.

> > docker -H 10.192.89.142 pull redis
> > Using default tag: latest
> > Pulling from library/redis
> > 6a5a5368e0c2: Extracting [==================================================>] 51.35 MB/51.35 MB
> > a3ed95caeb02: Download complete 
> > 2f1103ce5ca9: Download complete 
> > 086a40c85e01: Download complete 
> > 9a5e9d112ec4: Download complete 
> > dadc4b601bb4: Download complete 
> > b8066982e7a1: Download complete 
> > 2bcdfa1b63bf: Download complete 
> > Failed to write to image store: [POST /storage/{store_name}][500] WriteImage default  &{Code:0xc420ef7858 Message:NotAuthenticated}
> > 
> > docker  -H 10.192.89.142 run -it --name buz6  busybox
> > docker: Error response from daemon: Server error from portlayer: [POST /scopes/{scope}/containers][500] addContainerInternalServerError  &{Code:<nil> Message:NotAuthenticated}.
> > See 'docker run --help'.

The host has just one nic
[container-logs.zip](https://github.com/vmware/vic/files/542959/container-logs.zip)
screen shot 2016-10-20 at 2 32 15 pm

### PR DESCRIPTION
**Pull request by chengwang86**
Fixes #2835 

```
time="2016-10-28T20:15:54Z" level=debug msg="Error configuring vSphere Event Collector: Post https://sdkTunnel:8089/sdk: EOF" 
time="2016-10-28T20:16:59Z" level=warning msg="session keepalive error: Post https://sdkTunnel:8089/sdk: Service Unavailable (Error connecting tunnel on TCP socket: (null))" 
time="2016-10-28T20:21:59Z" level=warning msg="session keepalive error: ServerFaultCode: The session is not authenticated." 
time="2016-10-28T20:21:59Z" level=info msg="session keepalive re-authenticated" 
```

### CODE REVIEW

File: pkg/vsphere/session/session.go [hunk 1]
```diff
@@ -205,7 +206,26 @@ func (s *Session) Connect(ctx context.Context) (*Session, error) {
 
 	if s.Keepalive != 0 {
 		// TODO: add login() to the keep alive handler
-		vimClient.RoundTripper = session.KeepAlive(soapClient, s.Keepalive)
+		// vimClient.RoundTripper = session.KeepAlive(soapClient, s.Keepalive)
```
Comment (mdubya66) [inline pkg/vsphere/session/session.go]: remove the commented code :)

File: pkg/vsphere/session/session.go [hunk 2]
```diff
@@ -205,7 +206,25 @@ func (s *Session) Connect(ctx context.Context) (*Session, error) {
...
 		// TODO: add login() to the keep alive handler
-		vimClient.RoundTripper = session.KeepAlive(soapClient, s.Keepalive)
+		vimClient.RoundTripper = session.KeepAliveHandler(soapClient, s.Keepalive,
+			func(roundTripper soap.RoundTripper) error {
+				_, err := methods.GetCurrentTime(context.Background(), roundTripper)
+				if err != nil {
+					log.Warnf("session keepalive error: %s", err)
+
+					if isNotAuthenticated(err) {
+						err = login(ctx)
```
Comment (anchal-agrawal) [inline pkg/vsphere/session/session.go]: We can say

```
if err = login(ctx); err != nil {
  ...
} else {
  ...
}
```

**Comment by chengwang86** [comment]
@dougm @hickeng could you please review this? Thanks :)

**Comment by emlin** [comment]
minor comment. lgtm

File: pkg/vsphere/session/session.go [hunk 3]
```diff
@@ -205,7 +206,24 @@ func (s *Session) Connect(ctx context.Context) (*Session, error) {
 
 	if s.Keepalive != 0 {
 		// TODO: add login() to the keep alive handler
-		vimClient.RoundTripper = session.KeepAlive(soapClient, s.Keepalive)
+		vimClient.RoundTripper = session.KeepAliveHandler(soapClient, s.Keepalive,
+			func(roundTripper soap.RoundTripper) error {
+				_, err := methods.GetCurrentTime(context.Background(), roundTripper)
+				if err != nil {
```
Comment (hmahmood) [inline pkg/vsphere/session/session.go]:
```go
if err == nil {
  return nil
}
```
... and similar constructs below would get rid of the multiple indents.

**Comment by hmahmood** [comment]
lgtm

**Comment by dougm** [comment]
Using the bonneville code is ok for now, but we had wanted to find a way to avoid waiting s.Keepalive time to re-login.  Note that Keepalive currently defaults to 5min in the port layer.

We should add an integration test too.  The govc `session.ls` and `session.rm` commands were recently added with this test in mind.  We can remove an existing session, which would force login again.

**Comment by mdubya66** [comment]
+1 on integration tests but let's not hold up this merge for it.

**Comment by chengwang86** [comment]
ping

**Pull request merged**